### PR TITLE
USMO-259 - Support Java Async frameworks

### PR DESCRIPTION
**Pull request by val06**
### What does this PR do?

This PR adds support for handling ioctl messages from java asynchronous frameworks
We need to correlate on the kernel side between Encryption layer (SSLEngine) and the Transport layer (SocketChannel)

### Motivation

Asynchronous frameworks don't use blocking SSLSocket, hence we had to create a separate set of instrumentations to support it. By design, JDK makes a full separation between encryption and transport layers. Plain buffer is available only in the Encryption layer while the underlying connection information is accessible only by hooking the Transport layer

### Additional Notes

This PR was a straight forward implementation to support the async frameworks, further work can be done to improve efficiency

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/tls/java/erpc.h [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    if (0 != bpf_probe_read_user(&connection, sizeof(conn_tuple_t), bufferPtr)){
+        log_debug("[handle_connection_by_peer] failed to parse connection info for pid: %d\n", peer_key.pid);
+        return 1;
+    }
+    normalize_tuple(&connection);
+    bufferPtr+=sizeof(conn_tuple_t);
+
+    //read the host tuple (peer domain string and peer port)
+    if (0 != bpf_probe_read_user(&peer_key.peer, sizeof(peer_t), bufferPtr)){
+        log_debug("[handle_connection_by_peer] failed reading hostname location for pid %d\n", peer_key.pid);
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/java/erpc.h:99]: i'm not sure hostname location is the best description to the code, did I miss anything?

File: pkg/network/ebpf/c/protocols/tls/java/erpc.h [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+static int __always_inline is_usm_erpc_request(struct pt_regs *ctx) {
+    u32 cmd = PT_REGS_PARM3(ctx);
+    return cmd == USM_IOCTL_ID;
+}
+
+/*
+  handle_erpc_request ioctl request format :
+
+  struct {
+      u8           operation;  // REQUEST, CLOSE_CONNECTION, HOST, PLAIN
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/java/erpc.h:180]: REQUEST? PLAIN?

File: pkg/network/ebpf/c/protocols/tls/java/erpc.h
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/java/erpc.h:1]: general suggestion, it looks like we have code duplication in the different methods here, I do suggest to go over the functions and extract common code and use macro / functions for them.

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/tls/java/maps.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        // Kernels < 4.7.0 do not know about the per-cpu array map used
+        // in classification, preventing the program to load even though
+        // we won't use it. We change the type to a simple array map to
+        // circumvent that.
+        BPF_ARRAY_MAP(java_tls_hostname, __u32, 1)
+    #endif
+
+#else
+    BPF_PERCPU_ARRAY_MAP(java_tls_hostname, __u32, peer_domain_port_t, 1)
+#endif
```
[the remark this answers is not in the document]
Comment (val06) [reply pkg/network/ebpf/c/protocols/tls/java/maps.h]: what do you mean? i kinda copied that from kafka [implementation](https://github.com/DataDog/datadog-agent/blob/main/pkg/network/ebpf/c/protocols/kafka/maps.h#L10)

[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/tls/java/maps.h]: ignore that, it was an old comment from the first time I reviewed the PR

File: pkg/network/ebpf/c/protocols/tls/java/erpc.h
Comment (val06) [reply pkg/network/ebpf/c/protocols/tls/java/erpc.h:1]: do you have any examples? 
The only code block i could think of is reading the `conn_tuple_t` from the buffer which i felt doesn't deserve its own helper method.

Comment (guyarb) [reply pkg/network/ebpf/c/protocols/tls/java/erpc.h:1]: yes, talked about the conn_tuple_t reading, but it is negligible

**Review by guyarb — CHANGES_REQUESTED** [review]

**Review by guyarb — DISMISSED** [review]

**Review by guyarb — APPROVED** [review]

**Pull request merged**